overlord/snapshotstate/backend/backend.go

```diff
@@ -323,3 +346,116 @@ func addDirToZip(ctx context.Context, snapshot *client.Snapshot, w *zip.Writer,
+	comment := fmt.Sprintf("snapshot %d", id)
+
+	// prepare cache location to unpack the import file
+	p := path.Join(dirs.SnapCacheDir, "snapshots", string(id))
+	if _, err := os.Stat(p); !os.IsNotExist(err) {
+		return 0, nil, fmt.Errorf("snapshot import already in progress for ID `%d`", id)
+	}
+	if err := os.MkdirAll(p, 0755); err != nil {
+		return 0, nil, fmt.Errorf("failed creating import cache: %v", err)
+	}
+	defer os.RemoveAll(p)
+
+	// unpack the tar file to a temporary location (so the contents can be validated)
+	exportFound, size, err := unpackSnapshotImport(r, p)
+	if err != nil {
+		return 0, nil, err
+	}
+
+	if !exportFound {
+		return 0, nil, fmt.Errorf("snapshot import file incomplete: no export.json file")
+	}
+
+	// XXX: check the snapshot hashes (these are not in export.json at present)
+
+	// walk the cache directory to store the files
+	dir, err := osOpen(p)
+	if err != nil {
+		return 0, nil, fmt.Errorf("failed opening import cache: %v", comment)
+	}
+	defer dir.Close()
+	names, err := dirNames(dir, 100)
+	if err != nil {
+		return 0, nil, fmt.Errorf("failed read from import cache: %v", comment)
+	}
+
+	// move the files into place with the new local set ID
+	snapNames, err := moveCachedSnapshots(names, id, p)
+	return size, snapNames, err
+}
+
+func unpackSnapshotImport(r io.Reader, p string) (bool, int64, error) {
+	tr := tar.NewReader(r)
+	size := int64(0)
+	var tarErr error
+	var header *tar.Header
+	var exportFound bool
+	for tarErr == nil {
+		var skip bool
+		header, tarErr = tr.Next()
+		switch {
+		case tarErr == io.EOF:
+			skip = true
+		case tarErr != nil:
+			return false, 0, fmt.Errorf("failed reading snapshot import: %v", tarErr)
+		case header == nil:
+			// should not happen
+			skip = true
+		case header.Typeflag == tar.TypeDir:
+			// should not happen, but ignore directories
+			skip = true
+		}
+
+		if skip {
+			continue
+		}
+
+		if header.Name == "export.json" {
+			exportFound = true
+		}
+
+		target := path.Join(p, header.Name)
+		t, err := os.OpenFile(target, os.O_CREATE|os.O_RDWR, os.FileMode(header.Mode))
+		if err != nil {
+			return false, 0, fmt.Errorf("failed creating file `%s`: %v", target, err)
+		}
+		writtenSize, err := io.Copy(t, tr)
+		if err != nil {
+			return false, 0, fmt.Errorf("failed copying file `%s`: %v", target, err)
+		}
+		size += writtenSize
+
+		t.Close()
+	}
+	return exportFound, size, nil
+}
+
+func moveCachedSnapshots(names []string, id uint64, p string) ([]string, error) {
+	snaps := []string{}
+	for _, name := range names {
+		if name == "export.json" {
+			// ignore metadata file
+			continue
+		}
+		snapshot, err := snapshotFromFilename(name)
+		if err != nil {
+			return nil, err
+		}
+		snaps = append(snaps, snapshot.Snap)
+
+		// set the new setID and get the new filename
+		snapshot.SetID = id
+		new := Filename(snapshot)
+		old := path.Join(p, name)
+
+		if err := os.Rename(old, new); err != nil {
```

Comment:
Don't we need to update the set ID in the metadata as well? AFAICT right now it will carry the same ID that was set during export.

Reply:
That is set in the previous line:
```
		// set the new setID and get the new filename
		snapshot.SetID = id
		new := Filename(snapshot)
		old := path.Join(p, name)
```

Reply:
I meant the ID stored in the metadata file inside the snapshot itself. We import the snapshot and save if under a name that reflects the ID one request but AFAICT the metadata is not updated. From looking at other code, I think it will also break how snapshots are lited in `snap saved`.

Reply:
Yes, @bboozzoo is right. Unfortunately only the set id stored in the metadata inside the zip matters atm (updating it would require updating of the zip file). I'm working on changing that.

Reply:
This is now changed and the sid from the filename is now the authoritative part and the metadata is ignored.